### DISCUSSION TITLE
Include Resources with AvaloniaRuntimeXamlLoader.Parse()

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#9591] [Q&A] [answered]
With default settings the following code will not resolve application-level resources when parsing the XAML and creating the object tree.

```csharp
AvaloniaRuntimeXamlLoader.Parse<DataTemplate>("itemXaml")
```

Is there a way to pass existing resources or the application assembly to this or a similar method so resources can be used in this case?

Edit: Note I'm talking about more than one resource type:
 1. `avares:` URI references such as with `FontFamily="avares:\\"`
 2. `{StaticResource key}` and `{DynamicResource key}`

**Comment by maxkatz6** [comment, marked as answer]
This PR will allow passing IServiceProvider to the xaml loader: https://github.com/AvaloniaUI/Avalonia/pull/9541
IServiceProvider is used by markup extensions, like StaticResource. Here I added a test for this scenario: https://github.com/AvaloniaUI/Avalonia/blob/fix-parent-service-provider/tests/Avalonia.Markup.Xaml.UnitTests/Xaml/XamlIlTests.cs#L320
DynamicResource works differently IIRC, it uses dynamic visual tree instead of static XAML tree to resolve resources.

> avares: URI references such as with FontFamily="avares:\\"

That's different. But also it should be already possible to do.
Runtime XAML loader will try to find loaded assembly (remembering syntax of "avares:\\AssemblyName\Path.xaml").
If you use relative sources, Load method also has a "BaseUri" parameter, so it will know how to resolve an absolute URI.

That's if these resources are already available in the assembly.
If you want to load multiple XAML files that reference each other, use LoadGroup API:
https://github.com/AvaloniaUI/Avalonia/blob/fix-parent-service-provider/tests/Avalonia.Markup.Xaml.UnitTests/Xaml/StyleIncludeTests.cs#L57-L83

**Reply by robloo** [reply]
Thanks for the details.

> This PR will allow passing IServiceProvider to the xaml loader: https://github.com/AvaloniaUI/Avalonia/pull/9541
 IServiceProvider is used by markup extensions, like StaticResource.

It would be nice to have at least `{StaticResource}` extension support. Hoping that PR gets merged-in. You have a lot of PRs in limbo and I remember what happened to the icons :/

For now, I suppose the resources can just be loaded again from a file (duplicated) and passed to the LoadGroup API.

> DynamicResource works differently IIRC, it uses dynamic visual tree instead of static XAML tree to resolve resources.

So, it's not supported and there are no plans to do so? Less important but would be nice somehow.

>> avares: URI references such as with FontFamily="avares:\"
>
>That's different. But also it should be already possible to do.

Will test this one out specifically then. Probably missing a fully qualified assembly name some place.

> If you want to load multiple XAML files that reference each other, use LoadGroup API:

That's useful! Did not know about that.

**Reply by robloo** [reply]
>>>avares: URI references such as with FontFamily="avares:"
>>
>>That's different. But also it should be already possible to do.
>
>Will test this one out specifically then. Probably missing a fully qualified assembly name some place.

Quick update for anyone reading this in the future: Font's do work as indicated above. The `#Name` suffix (which isn't required on other XAML frameworks) was missing.

Code to build a FontFamily reference string (able to be passed into the runtime XAML loader as part of other XAML) may look like this:

```csharp
var symbolFont = Application.Current.GetResource<FontFamily>("AppSymbolFontFamily");
string fontResourceName = symbolFont.Key.Source.ToString() + "#Symbols";
```
